### PR TITLE
extend introspection with functions for array operations

### PR DESCRIPTION
**Pull request by dirk-thomas**
This builds on top of #22.

### CODE REVIEW

**Comment by wjwwood** [comment]
+1

File: rosidl_typesupport_introspection_cpp/resource/msg__type_support.cpp.template [hunk 1]
```diff
@@ -67,7 +103,18 @@ for index, field in enumerate(spec.fields):
     # bool is_upper_bound_
     print('    %s,  // is upper bound' % ('true' if field.type.is_upper_bound else 'false'))
     # unsigned long offset_
-    print('    offsetof(%s::%s::%s, %s)  // bytes offset in struct' % (spec.base_type.pkg_name, subfolder, spec.base_type.type, field.name))
+    print('    offsetof(%s::%s::%s, %s),  // bytes offset in struct' % (spec.base_type.pkg_name, subfolder, spec.base_type.type, field.name))
+    # void * default_value_
+    print('    nullptr,  // default value')  # TODO default value to be set
```
Comment (jackieokay) [inline rosidl_typesupport_introspection_cpp/resource/msg__type_support.cpp.template:120]: Is there an issue connected to this todo?

Comment (dirk-thomas) [inline rosidl_typesupport_introspection_cpp/resource/msg__type_support.cpp.template:120]: No.

File: rosidl_typesupport_introspection_cpp/resource/msg__type_support.cpp.template [hunk 2]
```diff
@@ -39,6 +39,42 @@ namespace rosidl_typesupport_introspection_cpp
...
+const void * get_function__@(spec.base_type.type)__@(field.name)(const void * untyped_member, size_t index)
+{
+@[if field.type.array_size and not field.type.is_upper_bound]@
+  const @(field.type.pkg_name)::msg::@(field.type.type) * member =
+    reinterpret_cast<const @(field.type.pkg_name)::msg::@(field.type.type) *>(untyped_member);
+@[else]@
+  const std::vector<@(field.type.pkg_name)::msg::@(field.type.type)> & member =
+    *reinterpret_cast<const std::vector<@(field.type.pkg_name)::msg::@(field.type.type)> *>(untyped_member);
+@[end if]@
+  return &member[index];
```
Comment (jackieokay) [inline rosidl_typesupport_introspection_cpp/resource/msg__type_support.cpp.template:64]: should you check if index is within the bounds of the vector, or do you expect to do error handling outside of this function?

Comment (dirk-thomas) [inline rosidl_typesupport_introspection_cpp/resource/msg__type_support.cpp.template:64]: The calling rmw implementation needs to make sure to only pass a valid index. It checks with `size_function` before or uses `resize_function` to increase the size as necessary.

Therefore I wouldn't add error handling in this function.

**Comment by jackieokay** [comment]
+1

**Comment by tfoote** [comment]
+1

**Comment by dirk-thomas** [comment]
I have added a second `get` function: one `const` the other `non-const`. This will allow `rmw_connext_dynamic_cpp` to be fixed to be const correct.

**Pull request merged**